storage/innobase/lock/lock0lock.cc

```diff
@@ -1636,21 +1761,38 @@ static void lock_rec_add_to_queue(unsigned type_mode, const hash_cell_t &cell,
 	if (type_mode & LOCK_WAIT) {
 		goto create;
 	} else if (lock_t *first_lock = lock_sys_t::get_first(cell, id)) {
+		bool bypass_mode= !is_supremum
+			&& lock_t::is_rec_exclusive_not_gap(type_mode);
+		bool has_s_lock_or_stronger= false;
 		for (lock_t* lock = first_lock;;) {
-			if (lock->is_waiting()
-			    && lock_rec_get_nth_bit(lock, heap_no)) {
-				goto create;
+			if (!lock_rec_get_nth_bit(lock, heap_no))
+				goto cont;
+			if (bypass_mode && lock_rec_can_be_bypassing(trx, lock))
+			{
+				has_s_lock_or_stronger= true;
 			}
+			/* There can be several locks suited for bypassing,
+			skip them all */
+			else if (lock->is_waiting() &&
+				 (!bypass_mode || !lock->can_be_bypassed(
+				    has_s_lock_or_stronger)))
+					goto create;
```

Comment:
`can_be_bypassed()` is checking `is_waiting()` too. Can we simplify this so that `is_waiting()` will be checked only once?